lib/linked_list.rb

```diff
@@ -12,120 +12,225 @@ def initialize(value, next_node = nil)
-      raise NotImplementedError
-    end
-
-    # method to print all the values in the linked list
-    def visit
-      raise NotImplementedError
-    end
-
-    # method to delete the first node found with specified value
-    def delete(value)
-      raise NotImplementedError
-    end
-
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    def reverse
-      raise NotImplementedError
-    end
-
-
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    def find_middle_value
-      raise NotImplementedError
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
+  
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  def add_first(value)
+    @head = Node.new(value, @head)
+  end
+  
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  def search(value)
+    current = @head
+    
+    until current.nil?
+      return true if current.data == value
+      current = current.next
+    end
+    
+    return false
+  end
+  
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    current = @head
+    return nil if current.nil?
+    max = 0
+    until current.nil?
+      if current.data > max
+        max = current.data
+      end
+      current = current.next
     end
-
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    def find_nth_from_end(n)
-      raise NotImplementedError
+    return max
+    
+  end
+  
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  def find_min
+    current = @head
+    return nil if current.nil?
+    min = current.data
+    
+    until current.nil?
+      if current.data < min
+        min = current.data
+      end
+      current = current.next
     end
-
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    def has_cycle
-      raise NotImplementedError
+    return min
+  end
+  
+  
+  # method that returns the length of the singly linked list
+  def length
+    count = 0
+    current = @head
+    until current.nil?
+      current = current.next
+      count +=1
+    end
+    return count
+  end
+  
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  def get_at_index(index)
+    current = @head
+    if current.nil?
+      return nil
+    elsif index == 0
+      return current.data
+    else
+      index.times do
+        current = current.next

```

Comment:
What if the list is shorter than the given index.

```suggestion
        current = current.next  unless current.nil?
```